### PR TITLE
[Blocks] Preserve source markup in invalid blocks

### PR DESCRIPTION
**Pull request by dmsnell**
## Status

Currently this is **ON HOLD** as I prepare a broader change that addresses the shortcomings noted in the discussion, namely the `originalContent` field and all the little places we need to compare the preserved markup with the re-generated markup, such as in the block comparer during a "Resolve" for an invalid block.

 - [x] Can we avoid updating all the test fixtures? I feel really bad about making an otherwise small PR blow up with changes that are hard to review. I considered modifying our custom snapshot testing runner to exclude the `sourceMarkup` attribute but wanted to hear thoughts first about it. I wondered if we rely on these fixtures and if they are providing much value to us. If they are, I wonder why we have a custom runner instead of using `jest`, if that's a carry-over from earlier times.
      - I've updated the test runner to ignore/remove the `sourceMarkup` property when comparing. On one hand I like this but on the other hand maybe it would be best to leave it in and update the fixtures with the 600+ lines of code that adds to this PR because it would be less surprising to come across. Any thoughts are appreciated.
 - [x] Identify places that are still inconsistent or already consistent. For example, the "Resolve" screen comparing the invalid block against its fix gets things right somehow. Other places might use `originalContent` and I'd like to audit all those places and even get rid of `originalContent` if we can.
     - [ ] [install-button](https://github.com/WordPress/gutenberg/blob/1a8971ba51d1f69b718bf2f77261184b43950217/packages/block-directory/src/plugins/get-install-missing/install-button.js#L28-L31) parses `originalContent` but should we have it parse `sourceMarkup` instead?
     - [ ] [block-compare](https://github.com/WordPress/gutenberg/blob/1a8971ba51d1f69b718bf2f77261184b43950217/packages/block-editor/src/components/block-compare/index.js) shows a diff view of the `innerHTML` between a source block and the re-generated markup. showing the difference with `sourceMarkup` would show the delimiters and also the inner blocks. inner blocks could be noisy, but also essential. showing the delimiters could be confusing.
     - [ ] [getBlockInnerHTML](https://github.com/WordPress/gutenberg/blob/1a8971ba51d1f69b718bf2f77261184b43950217/packages/blocks/src/api/serializer.js#L280) returns `innerHTML` if it fails to validate or re-generate block output. can this wipe out inner blocks?
     - [ ] ⚠️ [validateBlock](https://github.com/WordPress/gutenberg/blob/1a8971ba51d1f69b718bf2f77261184b43950217/packages/blocks/src/api/validation/index.js#L732-L736) only looks at `innerHTML` which means that if we get different `innerBlocks` we'll accept it as valid.
     - [ ] ⚠️ [isValidBlockContent](https://github.com/WordPress/gutenberg/blob/1a8971ba51d1f69b718bf2f77261184b43950217/packages/blocks/src/api/validation/index.js#L764-L776) overwrites a block's `innerBlocks` with an empty array for validation, completely ignoring them.
 - [x] Assess more concretely the inner blocks problem whereas we want to be able to edit and modify recognized inner blocks for parent blocks that are invalid. This isn't totally pertinent to this PR so I don't want to hold it up on that account, but I'd like to better understand the implications and future work before merging.
     - for the sake of this PR I don't think we have to consider this problem. we're not making it worse and we're not getting in the way of fixing it. all we'll be doing is making sure that on save an invalid block leaves the original markup.

## Description

Previously we have been re-generating a block's save content when
we fail to properly validate that block. While this can work in many
cases it breaks down quickly because of the invalidly-parsed attributes.

Consider a list block whose `<ul>` tag has been removed. The block will
fail to validate and then also fail to read the content of the list items
because the CSS selector used to source the attribute won't work. The
attribute for the list's content will be empty and `getSaveContent` will
return an empty list, this despite the fact that the contents were largely
fine in the post HTML when loaded.

In this patch we're threading a new property into the block node called
`sourceMarkup` whose purpose is to preserve the original HTML in a post
so that we can choose to preserve that on save instead of corrupting or
losing content through the HTML re-generation.

With `sourceMarkup` we get a less-processed version of the source to
relay into save which should eliminate a common point of frustration:
saving a post erases content.

### Invalid list before
Screen Shot 2022-02-02 at 11 41 00 AM

Note that the "Resolve" dialog already showed the proper content, but in the editor it looks like an empty list.

### Invalid list after
Screen Shot 2022-02-02 at 11 38 57 AM

The editor shows the proper list items and will preserve them on save.

## Types of changes
 - Preserving original block attributes, HTML, and inner blocks when saving an invalid block vs. attempting to re-generate that content from the attributes which parsed but failed to validate.
 - Render existing broken content as HTML in editor for invalid blocks instead of showing broken re-generated content.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [ ] I've updated related schemas if appropriate.

### CODE REVIEW

File: packages/blocks/src/api/parser/index.js
Comment (youknowriad) [inline packages/blocks/src/api/parser/index.js:140]: What's the reasoning for the removal of the condition here? Is it something that never happens?

File: packages/blocks/src/api/parser/index.js
Comment (youknowriad) [inline packages/blocks/src/api/parser/index.js:233]: It's not clear to me what's the difference between sourceMarkup and originalContent. As far as I remember, originalContent was introduced for the same reasons.

File: test/integration/full-content/full-content.test.js
Comment (youknowriad) [inline test/integration/full-content/full-content.test.js]: I think that's fine since we already have a dedicated snapshot for the source file, maybe we could also consider the same thing for originalContent.

File: packages/blocks/src/api/parser/index.js
Comment (dmsnell) [reply packages/blocks/src/api/parser/index.js:140]: previously we have been re-generating block output from the parsed inputs. this is the reason that the editor can silently corrupt or lose content. for HTML soup outside of the blocks though there's no way to re-generate the block since there is no block. because of this we special-case the HTML soup and use `innerHTML` directly, which ironically is closer to the correct behavior for regular blocks too.

after preserving the source markup every parsed block contains that original content which we can thread through and it doesn't matter if it's a block proper or not, so no more need for the conditional and special-casing.

File: packages/blocks/src/api/parser/index.js
Comment (dmsnell) [reply packages/blocks/src/api/parser/index.js:233]: The introduction of `originalContent` in #14443 was ironically made to support inner blocks even though we've got a few years to reflect on it and [confirm](https://github.com/WordPress/gutenberg/pull/14443#issuecomment-478716522) it breaks them. 🙃

I find `originalContent` as a name confusing particularly because it stands in sharp contrast to how we define "content" in `innerContent`. That is, `innerContent` is an array that specifically represents the view of the document including the inner blocks while `originalContent` is a string that specifically represents the view where the inner blocks are missing.

I considered removing `originalContent` altogether but I'm worried about what might be using it. As noted in the original PR we should be moving towards removing all use of `innerHTML` everywhere as it's an historical artifact going back to an original design question revolving around how to represent nesting.

`sourceMarkup` though has a clear referent that's not muddied by the confusing terminology (confusing not only because it means the opposite of what you should infer based on the block API but also because it implies it came from the original document when in fact it was re-generated on the fly from possibly faulty inputs).

With this change in we should be able to start removing references to `originalContent` and if possible, I'd rather do that in other work since this PR is fixing a longstanding issue and I'm nervous that if we start updating and changing other parts we up the risk of introducing regressions.

File: test/integration/full-content/full-content.test.js
Comment (dmsnell) [reply test/integration/full-content/full-content.test.js]: this sounds great. in fact I think we can go one step further and more positively talk about the attributes we want to check instead of removing the ones we don't. I'll post an update with what I mean.

Comment (dmsnell) [reply test/integration/full-content/full-content.test.js]: In [3a9ffa6](https://github.com/WordPress/gutenberg/pull/38460/commits/3a9ffa6e4f5e7772e8f91e3150dcefc403e0ffd0) I've adjusted this function but left `originalContent` for the same reason I didn't want to introduce the `sourceMarkup` - it turns this 30-line PR into a 700+ line PR. I wouldn't mind doing a follow-up PR though to update all the fixtures to remove the things we don't want and where that testing-methodology change could be reviewed by itself.

File: packages/blocks/src/api/parser/index.js
Comment (youknowriad) [reply packages/blocks/src/api/parser/index.js:233]: Ok, so the intent is indeed to change the way we compute `originalContent` and the name change is there to bring clarity. 

I think bringing clarity is a great argument but I wonder if we should just continue using `originalContent` that way we're not stuck supporting this old property forever and add some useless memory consumption to our stores. I'm hesitant personally here.

Comment (youknowriad) [reply packages/blocks/src/api/parser/index.js:233]: If we really want to stick with the renaming, we could potentially make `originalContent` a getter/setter property in order to add a `deprecated` call maybe.

**Review by youknowriad — COMMENTED** [review]
This is looking good overall for me despite my hesitation about the renaming.

Would love some 👀 from @mcsf here.

File: packages/blocks/src/api/parser/index.js
Comment (dmsnell) [reply packages/blocks/src/api/parser/index.js:233]: if you're uncomfortable keeping both I'd be in favor of removing `originalContent` entirely even at risk of breaking things. I'll check the usage in the missing block and in other places. `innerHTML` should go away too but I'm not going to change that here.

**Comment by dmsnell** [comment]
Rebuilt this after the merge of #38638, but I'm still trying to totally eliminate `originalContent` in another branch.

**Comment by dmsnell** [comment]
Unfortunately as I continue to try and eliminate `originalContent` I keep finding more and more places where the size of the change blows up. I'm really uncomfortable making such a huge change all at once, figuring we'll certainly miss a lot of regressions. I'm not sure we wouldn't still need `sourceMarkup` in the end anyway since the current loading process makes changes to the original block through things like legacy `autop` behaviors.

Tossing it out there that it might be practical to fix the existing problem (the editor wipes out or corrupts content from unrelated changes) with the small fix and then start chiseling away at `originalContent` and `innerHTML` which take a lot more effort and care. Having both values in place at the same time also means we can piecemeal break apart those `originalContent` and `innerHTML` uses instead of doing it all at once.

**Review by youknowriad — APPROVED** [review]
LGTM but I'd appreciate a second opinion here.

**Comment by dmsnell** [comment]
I'm going to wait for another opinion as well. Thank you @youknowriad 

The more I think about this though the more I think a different interface is going to be better. Possibly that's just two properties vs. one: `innerMarkup` and `outerMarkup`. Replacing `originalContent` is hard and many places (like the block comparison dialog) rely on having the HTML inside the comment delimiters. This is the difference between `originalContent` and `originalUndelimitedContent` except that those values break inner blocks.

This being said, I'm still comfortable pushing this out as-is since it's so small and because it does solve one major problem at hand and if we improve it afterwards and can remove `originalContent` then we can replace `sourceMarkup` or change it without changing the behaviors this introduces/fixes.

**Comment by dmsnell** [comment]
@youknowriad I'm closing this in favor of #38923 because I believe it provides a better longer-term solution with its `source` property, containing the original `rawBlock`, and doesn't require that we assess the same performance questions `sourceMarkup` does. Also, `sourceMarkup` just isn't sufficient on its own to handle all the cases we need for general broken-block preservation.

**Pull request closed without merging**
